src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CourseActivity.kt

```diff
@@ -0,0 +1,220 @@
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.AgencyLocationRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.CourseActivityRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.ReferenceCodeRepository
+import java.time.LocalDate
+
+@DslMarker
+annotation class CourseActivityDslMarker
+
+@TestDataDslMarker
+interface CourseActivityDsl {
+  @CourseActivityPayRateDslMarker
+  fun payRate(
+    iepLevelCode: String = "STD",
+    payBandCode: String = "5",
+    startDate: String = "2022-10-31",
+    endDate: String? = null,
+    halfDayRate: Double = 3.2,
+  ): CourseActivityPayRate
+
+  @CourseScheduleDslMarker
+  fun courseSchedule(
+    courseScheduleId: Long = 0,
+    scheduleDate: String = "2022-11-01",
+    startTime: String = "08:00",
+    endTime: String = "11:00",
+    slotCategory: SlotCategory = SlotCategory.AM,
+    scheduleStatus: String = "SCH",
+  ): CourseSchedule
+
+  @CourseScheduleRuleDslMarker
+  fun courseScheduleRule(
+    id: Long = 0,
+    startTimeHours: Int = 9,
+    startTimeMinutes: Int = 30,
+    endTimeHours: Int = 12,
+    endTimeMinutes: Int = 30,
+    monday: Boolean = true,
+    tuesday: Boolean = true,
+    wednesday: Boolean = true,
+    thursday: Boolean = true,
+    friday: Boolean = true,
+    saturday: Boolean = false,
+    sunday: Boolean = false,
+  ): CourseScheduleRule
+}
+
+@Component
+class CourseActivityBuilderRepository(
+  private val courseActivityRepository: CourseActivityRepository? = null,
+  private val agencyLocationRepository: AgencyLocationRepository? = null,
+  private val iepLevelRepository: ReferenceCodeRepository<IEPLevel>? = null,
+  private val agencyInternalLocationRepository: AgencyInternalLocationRepository? = null,
+) {
+  fun save(courseActivity: CourseActivity) = courseActivityRepository?.save(courseActivity)
+  fun lookupAgency(id: String): AgencyLocation = agencyLocationRepository?.findByIdOrNull(id)!!
+  fun lookupIepLevel(code: String): IEPLevel =
+    iepLevelRepository?.findByIdOrNull(ReferenceCode.Pk(IEPLevel.IEP_LEVEL, code))!!
+  fun lookupAgencyInternalLocation(locationId: Long): AgencyInternalLocation? =
+    agencyInternalLocationRepository?.findByIdOrNull(locationId)
+}
+
+@Component
+class CourseActivityBuilderFactory(
+  private val repository: CourseActivityBuilderRepository? = null,
+  private val courseActivityPayRateBuilderFactory: CourseActivityPayRateBuilderFactory = CourseActivityPayRateBuilderFactory(),
+  private val courseScheduleBuilderFactory: CourseScheduleBuilderFactory = CourseScheduleBuilderFactory(),
+  private val courseScheduleRuleBuilderFactory: CourseScheduleRuleBuilderFactory = CourseScheduleRuleBuilderFactory(),
+) {
+  fun builder(): CourseActivityBuilder {
+    return CourseActivityBuilder(
+      repository,
+      courseActivityPayRateBuilderFactory,
+      courseScheduleBuilderFactory,
+      courseScheduleRuleBuilderFactory,
+    )
+  }
+}
```

Comment:
The only child of ProgramService is CourseActivity. Its builder factory needs:
* to save itself (in the repository)
* to lookup any dependent entities (such as reference codes), also in the repository
* builder factories for any of its children